### PR TITLE
Add support for error.culprit

### LINKED ISSUE #192: Support for error.culprit
This would be the first in_app frame of the deepest nested exception

### PR DESCRIPTION
**Pull request by felixbarny**
closes elastic/apm-agent-java#192

### CODE REVIEW

**Comment by felixbarny** [comment]
screen shot 2018-09-17 at 14 29 30

File: apm-agent-core/src/main/java/co/elastic/apm/impl/error/ErrorCapture.java [hunk 1]
```diff
@@ -131,4 +135,41 @@ public TraceContext getTraceContext() {
...
+    }
+
+    private void setCulprit(StackTraceElement stackTraceElement) {
+        culprit.append(stackTraceElement.getClassName())
+            .append('.')
+            .append(stackTraceElement.getMethodName())
+            .append('(')
+            .append(stackTraceElement.getFileName())
+            .append(':')
+            .append(stackTraceElement.getLineNumber())
```
[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/impl/error/ErrorCapture.java]: Actually, why not using `StackTraceElement#toString`?

File: apm-agent-core/src/main/java/co/elastic/apm/impl/ElasticApmTracer.java [hunk 2]
```diff
@@ -217,7 +217,7 @@ public void captureException(@Nullable Throwable e) {
 
     public void captureException(long epochTimestampMillis, @Nullable Throwable e, @Nullable AbstractSpan<?> active) {
         if (e != null) {
-            ErrorCapture error = new ErrorCapture();
+            ErrorCapture error = errorPool.createInstance();
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/impl/ElasticApmTracer.java]: both, actually :D 
but not necessarily related to this change

File: apm-agent-core/src/main/java/co/elastic/apm/impl/error/ErrorCapture.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/impl/error/ErrorCapture.java]: to avoid allocations

**Comment by felixbarny** [comment]
I think it would be difficult to come up with such a hash because the culprit might be a totally different method even if the class and the message is identical. For example, a NumberFormatException occurring in two different methods. I think as soon as the APM server has support for nested exceptions, it should compute the culprit to save some computation time in the agents.

**Comment by felixbarny** [comment]
But I think it's a good idea to throttle the number of captured errors in general. I would not restrict it to throttle the computation of `culprit` but rather ignore errors after a certain threshold, like 100 per minute.

**Pull request merged**